cf-serverd/server_tls.c

```diff
@@ -404,108 +404,139 @@ static int ServerSendWelcome(const ServerConnectionState *conn)
+    if (conn->conn_info->status == CONNECTIONINFO_STATUS_ESTABLISHED)
+    {
+        return 1;
+    }
+    assert(ConnectionInfoSSL(conn->conn_info) == NULL);
+    SSL *ssl = SSL_new(SSLSERVERCONTEXT);
+    if (ssl == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "SSL_new: %s",
+            TLSErrorString(ERR_get_error()));
+        return -1;
+    }
+    ConnectionInfoSetSSL(conn->conn_info, ssl);
+
+    /* Pass conn_info inside the ssl struct for TLSVerifyCallback(). */
+    SSL_set_ex_data(ssl, CONNECTIONINFO_SSL_IDX, conn->conn_info);
+
+    /* Now we are letting OpenSSL take over the open socket. */
+    SSL_set_fd(ssl, ConnectionInfoSocket(conn->conn_info));
+
+    int ret = SSL_accept(ssl);
+    if (ret <= 0)
+    {
+        TLSLogError(ssl, LOG_LEVEL_ERR,
+                    "Failed to accept TLS connection", ret);
+        return -1;
+    }
+
+    Log(LOG_LEVEL_VERBOSE, "TLS version negotiated: %8s; Cipher: %s,%s",
+        SSL_get_version(ssl),
+        SSL_get_cipher_name(ssl),
+        SSL_get_cipher_version(ssl));
+
+    return 1;
+}
 
 /**
  * @brief Accept a TLS connection and authenticate and identify.
+ *
+ * This function uses trustkeys to trust new keys and updates lastseen
+ *
+ * @see BasicServerTLSSessionEstablish
  * @note Various fields in #conn are set, like username and keyhash.
+ * @return 1 for success -1 for error
  */
 int ServerTLSSessionEstablish(ServerConnectionState *conn)
 {
-    if (conn->conn_info->status != CONNECTIONINFO_STATUS_ESTABLISHED)
+    if (conn->conn_info->status == CONNECTIONINFO_STATUS_ESTABLISHED)
     {
-        assert(ConnectionInfoSSL(conn->conn_info) == NULL);
-        SSL *ssl = SSL_new(SSLSERVERCONTEXT);
-        if (ssl == NULL)
-        {
-            Log(LOG_LEVEL_ERR, "SSL_new: %s",
-                TLSErrorString(ERR_get_error()));
-            return -1;
-        }
-        ConnectionInfoSetSSL(conn->conn_info, ssl);
+        return 1;
+    }
 
-        /* Pass conn_info inside the ssl struct for TLSVerifyCallback(). */
-        SSL_set_ex_data(ssl, CONNECTIONINFO_SSL_IDX, conn->conn_info);
+    int ret = BasicServerTLSSessionEstablish(conn);
+    assert(ret == 1 || ret == -1); // 1 == success, -1 == error
+    if (ret == -1)
+    {
+        return -1;
+    }
 
-        /* Now we are letting OpenSSL take over the open socket. */
-        SSL_set_fd(ssl, ConnectionInfoSocket(conn->conn_info));
+    Log(LOG_LEVEL_VERBOSE, "TLS session established, checking trust...");
 
-        int ret = SSL_accept(ssl);
-        if (ret <= 0)
-        {
-            TLSLogError(ssl, LOG_LEVEL_ERR,
-                        "Failed to accept TLS connection", ret);
-            return -1;
-        }
+    /* Send/Receive "CFE_v%d" version string, agree on version, receive
+       identity (username) of peer. */
+    char username[sizeof(conn->username)] = "";
+    bool b = ServerIdentificationDialog(conn->conn_info,
+                                        username, sizeof(username));
+    if (b != true)
```

Comment:
This is copied directly from existing code. Also:

> Explicit comparisons are better than implicit

We do a lot of things that are unnecessary for clarity. I would do one of:
```
if ( b == false )

Reply:
Yeah, renaming the variable makes sense. Because we also have this:
> * Explicit comparisons are better than implicit, i.e. prefer writing                                                                                                                               
  ```if (number == 0)``` or ```if (pointer == NULL)```                                                                                                                                             
  instead of ~~```if (!number)```~~ or ~~```if (!pointer)```~~. It only makes                                                                                                                      
  sense to test booleans directly, for example ```if (is_valid)``` is good.